### PR TITLE
Fix errors when declaring fixed array types using non-const sizes

### LINKED ISSUE #41938: [Bug]: Invalid error and compiler crash when using a configurable variable to specify length of a top level const array
### Description

As per spec to define the length of an array an integer literal or a constant expression should be used. (https://ballerina.io/spec/lang/master/#array-type-descriptor). In the case of using a configurable variable to define the length of an const array in top level (invalid case) a compiler crash occurs and an invalid error is given. This should produce the error ` invalid reference expression '[name]' as array size: expected a constant reference expression
error: compilation contains errors`.

### Steps to Reproduce

```ballerina
configurable int length = 1;

const int[length] y = [1];

public function main() {

}
```

### Affected Version(s)

2201.8.4

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #41938 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 1]
```diff
@@ -2015,3 +2015,6 @@ error.inferred.query.construct.type.as.stream=\
 
 error.invalid.binding.pattern.in.on.fail=\
   invalid binding pattern in ''on fail'' clause: only a capture binding pattern or an error binding pattern is allowed
+
+error.undefined.constant.symbol=\
+  undefined constant symbol ''{0}''
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:2020]: Is it necessary to define a new error?

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:2020]: I guess yes. Similar errors related to this issue include `undefined symbol` and `expression is not a constant expression`. There are two distinct scenarios to address:

1. In the first case, there exists a symbol in the scope, but it is not a constant.
2. In the second case, there is no symbol in the scope.

Using the `undefined symbol` error would be inaccurate in the first case. On the other hand, employing the `expression is not a constant expression` error is not feasible as it includes a broader range of expressions.
[const-expr defined in the spec](https://ballerina.io/spec/lang/master/#section_6.4)

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:2020]: Variables and constants share the [same symbol space](https://ballerina.io/spec/lang/master/#section_5.9.3), and therefore, using a single `undefined symbol` error is OK for when a symbol is undefined, IMO.

Also, the requirements for constant expressions are specific, and I believe it is better to give the exact error.

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:2020]: ACK. Let's keep a new error for this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -4364,11 +4364,13 @@ void analyzeNode(BLangNode node, BType expType, AnalyzerData data) {
...
         if (constant.typeNode != null && !types.isAllowedConstantType(constant.typeNode.getBType())) {
+            boolean isNeverType = types.isNeverTypeOrStructureTypeWithARequiredNeverMember(
+                    constant.typeNode.getBType());
             if (types.isAssignable(constant.typeNode.getBType(), symTable.anydataType) &&
-                    !types.isNeverTypeOrStructureTypeWithARequiredNeverMember(constant.typeNode.getBType())) {
+                    !isNeverType) {
                 dlog.error(constant.typeNode.pos, DiagnosticErrorCode.CONSTANT_DECLARATION_NOT_YET_SUPPORTED,
                         constant.typeNode);
-            } else {
+            } else if (isNeverType) {
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:4373]: Do not we need to handle `else` case?

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 1 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:2020]: > Variables and constants share the [same symbol space](https://ballerina.io/spec/lang/master/#section_5.9.3), and therefore, using a single `undefined symbol` error is OK for when a symbol is undefined, IMO.
> 
> Also, the requirements for constant expressions are specific, and I believe it is better to give the exact error.

+1, IMO this error msg is not correct. We should show a better one that says the expression isn't a constant. By seeing this error message, if someone tries to declare another constant with the same variable, they get an error saying the symbol is redeclared.

**Comment by chiranSachintha** [comment]
```ballerina
const int[length] y = [1];
const int length = 1;
public function main() {

}
```
If we change the order as above example, we're getting the following error. This is not correct. We need to fix this as well, If not this become breaking change in future releases.
```
ERROR [test.bal:(16:7,16:18)] undefined constant symbol 'length'
ERROR [test.bal:(16:23,16:26)] incompatible types: expected 'other', found '[1]'
```

**Comment by chiranSachintha** [comment]
There are few other scenarios that are currently not functioning correctly.

Case 1:
```ballerina
int[length] y = [1];
int length = 1; // undefined symbol 'length'

```

Case 2:
```ballerina
const int[length] y = [1];
const int length = 1;
```

If we consider the following scenarios, they are functioning correctly by giving an error: "invalid reference expression 'length' as array size: expected a constant reference expression."

Case 1:
```ballerina
int length = 1;

public function main() {
    int[length] y = [1];
}
```

Case 2:
```ballerina
int length = 1;
int[length] y = [1];
```

For the scenarios that are not working properly, it is better if we can give same error as "invalid reference expression 'length' as array size: expected a constant reference expression."

For that IMO we need different approach to fix this. In the current implementation, we are trying to resolve the type node of top level nodes (global variables and constants) before defining the all the symbols in the top level node. 

A better way to address this issue would be:

1. First, define all the symbols for the top-level nodes without resolving their type nodes.
2. Then, when visiting the semantic analyzer, resolve the type nodes.(While resolving type node, we can check whether that the expression is constant or not. This logic is already present in the symbol resolver)

**Comment by LakshanWeerasinghe** [comment]
> ```ballerina
> const int[length] y = [1];
> const int length = 1;
> public function main() {
> 
> }
> ```
> 
> If we change the order as above example, we're getting the following error. This is not correct. We need to fix this as well, If not this become breaking change in future releases.
> 
> ```
> ERROR [test.bal:(16:7,16:18)] undefined constant symbol 'length'
> ERROR [test.bal:(16:23,16:26)] incompatible types: expected 'other', found '[1]'
> ```

This case is addressed in #41978.

**Comment by prakanth97** [comment]
> There are few other scenarios that are currently not functioning correctly.
> 
> Case 1:
> 
> ```ballerina
> int[length] y = [1];
> int length = 1; // undefined symbol 'length'
> ```
> 
> Case 2:
> 
> ```ballerina
> const int[length] y = [1];
> const int length = 1;
> ```
> 
> If we consider the following scenarios, they are functioning correctly by giving an error: "invalid reference expression 'length' as array size: expected a constant reference expression."
> 
> Case 1:
> 
> ```ballerina
> int length = 1;
> 
> public function main() {
>     int[length] y = [1];
> }
> ```
> 
> Case 2:
> 
> ```ballerina
> int length = 1;
> int[length] y = [1];
> ```
> 
> For the scenarios that are not working properly, it is better if we can give same error as "invalid reference expression 'length' as array size: expected a constant reference expression."
> 
> For that IMO we need different approach to fix this. In the current implementation, we are trying to resolve the type node of top level nodes (global variables and constants) before defining the all the symbols in the top level node.
> 
> A better way to address this issue would be:
> 
> 1. First, define all the symbols for the top-level nodes without resolving their type nodes.
> 2. Then, when visiting the semantic analyzer, resolve the type nodes.(While resolving type node, we can check whether that the expression is constant or not. This logic is already present in the symbol resolver)
> 
> Currently, we are resolve only the types and constants through the depth-first algorithm. We need to resolve global variables as well.

To resolve this issue properly, we have to resolve the `global variables` and `xmlns decl` as well with depth-first approach (global variable and xmlns decl are excluded from adding them to the modTable which are also on the same scope). 

Take this below case,
```
const int[length] y = [1];
int length = 1;
```
1. To resolve the `y` then we resolve the type descriptor which is `int[length]`.
2. When resolving the type desc, we do lookup in the scope, if it `length` define if it is not, then define global variable `length`.
3. Check `length` is `cons-expr` or not, log the error something like `lenght is not an const expression`

**Pull request closed without merging**